File: lib/ledger_dsl/costed_amount.rb
# frozen_string_literal: true
# typed: strict

require 'sorbet-runtime'
require 'bigdecimal'

require_relative 'amount'
require_relative 'unit'

# DSL for creating financial journals
module LedgerDSL
  # Represents an amount with an optional associated cost
  class CostedAmount
    extend T::Sig

    # Type of cost
    class CostType < T::Enum
      enums do
        None = new
        Total = new
        Unit = new
      end
    end

    sig do
      params(
        amount: T.any(Amount, BigDecimal, Integer, Float, Rational, String),
        cost: T.nilable(ToAmount),
        type: CostType
      ).void
    end
    def initialize(amount, cost: nil, type: CostType::None)
      # Extract cost and type from a string amount
      if cost.nil? && amount.is_a?(String)
        match = amount.match(/(.*)(@@?)(.*)/)
        if match
          amount = T.must(match[1]).strip
          cost = T.must(match[3]).strip
          type = match[2] == '@' ? CostType::Unit : CostType::Total
        end
      end

      @amount = T.let(Amount.from(amount), Amount)
      @cost = T.let(cost && Amount.from(cost), T.nilable(Amount))
      @type = T.let(type, CostType)
    end

    sig { params(costed_amount: ToCostedAmount).returns(CostedAmount) }
    def self.from(costed_amount)
      return costed_amount if costed_amount.is_a?(CostedAmount)

      CostedAmount.new(costed_amount)
    end

    sig { returns(String) }
    def to_ledger
      amount = @amount.to_ledger
      if @cost
        cost = @cost.to_ledger
        if @type == CostType::Unit
          "#{amount} @ #{cost}"
        else
          "#{amount} @@ #{cost}"
        end
      else
        amount
      end
    end

    sig { params(amount: ToAmount, unit: Unit).returns(CostedAmount) }
    def total_cost(amount, unit = EMPTY_UNIT)
      @cost = Amount.from(amount, unit)
      @type = CostType::Total
      self
    end

    sig { params(amount: ToAmount, unit: Unit).returns(CostedAmount) }
    def unit_cost(amount, unit = EMPTY_UNIT)
      @cost = Amount.from(amount, unit)
      @type = CostType::Unit
      self
    end
  end

  ToCostedAmount = T.type_alias { T.any(CostedAmount, Amount, BigDecimal, Integer, Float, Rational, String) }
end

File: lib/entities/costed_amount.ts
import { Amount, ToAmount, ToValue } from "./amount";
import { EMPTY_UNIT, ToUnit } from "./unit";

/**
 * Type of cost.
 */
export enum CostType {
  None = "None",
  Total = "Total",
  Unit = "Unit",
}

/**
 * Types that can be converted to a costed amount.
 */
export type ToCostedAmount = CostedAmount | ToAmount;

/**
 * Represents an amount with an optional associated cost.
 */
export class CostedAmount {
  static COSTED_AMOUNT_REGEX = /^(.*)(@@?)(.*)$/;

  #amount: Amount;
  #type: CostType;
  #cost: Amount | null;

  /**
   * @param costedAmount - The costed amount to copy.
   */
  constructor(costedAmount: CostedAmount);
  /**
   * @param amount - The base amount.
   * @param cost - The optional cost.
   * @param type - The type of cost. Default is CostType.None.
   */
  constructor(amount: ToAmount, cost?: ToAmount | null, type?: CostType);
  /**
   * @param toCostedAmount - The value to convert to a costed amount.
   */
  constructor(toCostedAmount: ToCostedAmount);
  constructor(
    a: ToCostedAmount,
    b: ToAmount | null = null,
    c: CostType | null = null,
  ) {
    // Clone another costed amount.
    if (a instanceof CostedAmount) {
      this.#amount = a.#amount;
      this.#type = a.#type;
      this.#cost = a.#cost;
      return;
    }

    // Extract cost and type from a string amount
    if (typeof a === "string" && b === null && c === null) {
      const match = a.match(CostedAmount.COSTED_AMOUNT_REGEX);
      if (match) {
        this.#amount = new Amount(match[1]!.trim());
        this.#type = match[2] === "@" ? CostType.Unit : CostType.Total;
        this.#cost = new Amount(match[3]!.trim());
        return;
      }
    }

    this.#amount = new Amount(a);
    this.#type = c ?? CostType.None;
    this.#cost = b === null ? null : new Amount(b);
  }

  /**
   * Gets the base amount.
   */
  public get amount(): Amount {
    return this.#amount;
  }

  /**
   * Sets the base amount.
   */
  public set amount(amount: ToAmount) {
    this.#amount = new Amount(amount);
  }

  /**
   * Gets the cost amount, if any.
   */
  public get cost(): Amount | null {
    return this.#cost;
  }

  /**
   * Sets the cost amount.
   */
  public set cost(cost: ToAmount | null) {
    this.#cost = cost === null ? null : new Amount(cost);
  }

  /**
   * Gets the cost type.
   */
  public get type(): CostType {
    return this.#type;
  }

  /**
   * Sets the cost type.
   */
  public set type(type: CostType) {
    this.#type = type;
  }

  /**
   * Sets a total cost for this amount.
   *
   * @param amount - The cost amount.
   * @returns This CostedAmount for chaining.
   */
  public totalCost(amount: Amount): CostedAmount;
  /**
   * Sets a total cost for this amount.
   *
   * @param value - The cost value.
   * @param unit - The unit for the cost.
   * @returns This CostedAmount for chaining.
   */
  public totalCost(value: ToValue, unit: ToUnit): CostedAmount;
  public totalCost(a: ToAmount, b: ToUnit = EMPTY_UNIT): CostedAmount {
    this.#cost = a instanceof Amount ? a : new Amount(a, b);
    this.#type = CostType.Total;
    return this;
  }

  /**
   * Sets a per-unit cost for this amount.
   *
   * @param amount - The cost amount.
   */
  public unitCost(amount: Amount): CostedAmount;
  /**
   * Sets a per-unit cost for this amount.
   *
   * @param value - The cost value.
   * @param unit - The unit for the cost.
   * @returns This CostedAmount for chaining.
   */
  public unitCost(value: ToValue, unit: ToUnit): CostedAmount;
  public unitCost(a: ToAmount, b: ToUnit = EMPTY_UNIT): CostedAmount {
    this.#cost = a instanceof Amount ? a : new Amount(a, b);
    this.#type = CostType.Unit;
    return this;
  }

  /**
   * Converts the costed amount to a ledger-formatted string.
   *
   * @returns The formatted ledger string.
   */
  public toLedger(): string {
    const amount = this.#amount.toLedger();
    if (this.#cost) {
      const cost = this.#cost.toLedger();
      if (this.#type === CostType.Unit) {
        return `${amount} @ ${cost}`;
      } else {
        return `${amount} @@ ${cost}`;
      }
    } else {
      return amount;
    }
  }
}
